File: src/algorithms/python/H-Index.py
# -*- coding: utf-8 -*-


class Solution(object):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        citations.sort()
        for index in range(len(citations)):
            tmp = len(citations) - index
            if tmp <= citations[index]:
                if index == 0 or tmp >= citations[index - 1]:
                    return tmp
        return 0


if __name__ == "__main__":
    solution = Solution()
    print solution.hIndex([3, 0, 6, 1, 5])
    print solution.hIndex([100])

File: src/algorithms/cpp/H-Index.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int hIndex(vector<int>& citations) {
        sort(citations.begin(), citations.end());
        for (int i = SZ(citations); i >= 1; -- i) {
            int le_num = SZ(citations) - i;
            if (citations[le_num - 1] <= i && citations[le_num] >= i)
                return i;
        }
        return 0;
    }
};

int main() {
    return 0;
}
